### PR TITLE
Add information of hashAgg for memory control

### PR DESCRIPTION
**Pull request by wshwsh12**
### First-time contributors' checklist 

- [x] I've signed [**Contributor License Agreement**](https://cla-assistant.io/pingcap/docs-cn) that's required for repo owners to accept my contribution.

### Which TiDB version(s) do your changes apply to? (Required)

**Tips for choosing the affected version(s):**

By default, **CHOOSE MASTER ONLY** so your changes will be applied to the next TiDB major or minor releases. If your PR involves a product feature behavior change or a compatibility change, **CHOOSE THE AFFECTED RELEASE BRANCH(ES) AND MASTER**.

For details, see [tips for choosing the affected versions (in Chinese)](https://github.com/pingcap/docs-cn/blob/master/CONTRIBUTING.md#版本选择指南).

- [x] master (the latest development version)
- [ ] v5.1 (TiDB 5.1 versions)
- [ ] v5.0 (TiDB 5.0 versions)
- [ ] v4.0 (TiDB 4.0 versions)
- [ ] v3.1 (TiDB 3.1 versions)
- [ ] v3.0 (TiDB 3.0 versions)
- [ ] v2.1 (TiDB 2.1 versions)

### What is the related PR or file link(s)?

- This PR is translated from:
- Other reference link(s):

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

File: configure-memory-usage.md [hunk 1]
```diff
@@ -124,3 +124,50 @@ server-memory-quota = 34359738368
     * `record path`：表示状态文件存放的目录
 
 5. 通过访问状态文件所在目录（该示例中的目录为 `/tmp/1000_tidb/MC4wLjAuMDo0MDAwLzAuMC4wLjA6MTAwODA=/tmp-storage/record`），可以得到一组文件，其中包括 `goroutinue`、`heap`、`running_sql` 3 个文件，文件以记录状态文件的时间为后缀。这 3 个文件分别用来记录报警时的 goroutine 栈信息，堆内存使用状态，及正在运行的 SQL 信息。其中 `running_sql` 文件内的日志格式请参考 [`expensive-queries`](/identify-expensive-queries.md)。
+
+## 使用串行的 HashAgg 实现来缓解内存使用
+
+默认配置下，tidb-server 会使用并行的 HashAgg 实现来进行聚合函数的计算。
+但是并行的 HashAgg 实现有如下缺点：
```
Comment (TomShawn) [inline configure-memory-usage.md:131]:
```suggestion
默认配置下，tidb-server 会使用并行的 HashAgg 实现来进行聚合函数的计算。但是并行的 HashAgg 实现有如下缺点：
```

File: configure-memory-usage.md [hunk 2]
```diff
@@ -124,3 +124,49 @@ server-memory-quota = 34359738368
     * `record path`：表示状态文件存放的目录
 
 5. 通过访问状态文件所在目录（该示例中的目录为 `/tmp/1000_tidb/MC4wLjAuMDo0MDAwLzAuMC4wLjA6MTAwODA=/tmp-storage/record`），可以得到一组文件，其中包括 `goroutinue`、`heap`、`running_sql` 3 个文件，文件以记录状态文件的时间为后缀。这 3 个文件分别用来记录报警时的 goroutine 栈信息，堆内存使用状态，及正在运行的 SQL 信息。其中 `running_sql` 文件内的日志格式请参考 [`expensive-queries`](/identify-expensive-queries.md)。
+
+## 使用串行的 HashAgg 实现来缓解内存使用
```
Comment (XuHuaiyu) [inline configure-memory-usage.md:128]: 重新组织下这里的结构

```suggestion
## tidb-server 其它内存控制策略
### 流控
说一下流控是什么意思，对外表现是怎样的（比如日志里面会有什么信息）
### 落盘
- 说一下落盘受哪个变量控制，链接过去。
- 目前支持落盘的算子有哪些。
- hashagg 落盘是从哪个版本引入的。
- 文档中不用提及并行串行的概念。含 hashagg 的 sql 引起 OOM 时，可以尝试设置 concurrency = 1 来触发落盘。
- 再说下 hashagg 落盘的限制，1. 默认配置下 hashagg 无法进行落盘 2. 含 distinct 的 agg 无法进行落盘。
```

File: configure-memory-usage.md [hunk 3]
```diff
@@ -124,3 +124,49 @@ server-memory-quota = 34359738368
...
+1. 配置 SQL 的 Memory Quota 为 1GB（默认 1GB）。 `set tidb_mem_quota_query = 1 << 30`
+2. 创建单表 `CREATE TABLE t(a int);` 并插入 256 行不同的数据。
+3. 尝试执行 SQL `explain analyze select /*+ HASH_AGG() */ count(*) from t t1 join t t2 join t t3 group by t1.a, t2.a, t3.a;`。该 SQL 占用巨大的内存，会被 tidb-server cancel 掉。
+
+    ```sql
+    [tidb]> explain analyze select /*+ HASH_AGG() */ count(*) from t t1 join t t2 join t t3 group by t1.a, t2.a, t3.a;
+    ERROR 1105 (HY000): Out Of Memory Quota![conn_id=3]
+    ```
+
+4. 通过用户变量 `set tidb_executor_concurrency = 1;`，将执行器的并发度调整为 1。此时，HashAgg 会使用串行的执行逻辑。
```
Comment (Liuxiaozhen12) [inline configure-memory-usage.md:148]:
```suggestion
4. 通过设置用户变量，将执行器的并发度调整为 1：
   `set tidb_executor_concurrency = 1`
   此时，HashAgg 会使用串行的执行逻辑。
```

File: configure-memory-usage.md [hunk 4]
```diff
@@ -124,3 +124,49 @@ server-memory-quota = 34359738368
...
+5. 执行相同的 SQL，SQL 可以执行成功。可以从详细的执行计划看出，HashAgg 使用了 600MB 的硬盘空间。
```
Comment (Liuxiaozhen12) [inline configure-memory-usage.md:149]:
```suggestion
5. 执行相同的 SQL 语句，不再返回错误，可以执行成功。从详细的执行计划可以看出，HashAgg 使用了 600MB 的硬盘空间。
```

File: configure-memory-usage.md [hunk 5]
```diff
@@ -124,3 +124,49 @@ server-memory-quota = 34359738368
...
+    |     └─TableReader_17(Probe)     | 256.00      | 256      | root      |               | time:211.1µs, loops:2, cop_task: {num: 1, max: 295.5µs, proc_keys: 0, rpc_num: 1, rpc_time: 279.7µs, copr_cache_hit_ratio: 0.00}                                    | data:TableFullScan_16                                           | 885 Bytes | N/A      |
+    |       └─TableFullScan_16        | 256.00      | 256      | cop[tikv] | table:t1      | tikv_task:{time:71.4µs, loops:256}                                                                                                                                  | keep order:false, stats:pseudo                                  | N/A       | N/A      |
+    +---------------------------------+-------------+----------+-----------+---------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------+-----------------------------------------------------------------+-----------+----------+
+    9 rows in set (1 min 37.428 sec)
+    ```
+
+> **注意：**
+>
+> + HashAgg 目前仅串行的实现支持落盘功能。使用并行的 HashAgg 且内存占用过大时，无法进行落盘。
```
Comment (Liuxiaozhen12) [inline configure-memory-usage.md:171]:
```suggestion
> + HashAgg 目前仅支持串行时的落盘功能。使用并行的 HashAgg 且内存占用过大时，无法进行落盘。
```

**Review by Liuxiaozhen12 — COMMENTED** [review]
not sure the form of command I suggested is correct or not. Just suggestions.

File: configure-memory-usage.md [hunk 6]
```diff
@@ -124,3 +124,49 @@ server-memory-quota = 34359738368
...
+2. 无法进行落盘操作。当 SQL 的内存使用超过 Memory Quota 时，tidb-server 会 cancel 掉这条 SQL。
+
+在对 SQL 执行时间要求不严格的情况下，我们可以通过使用串行的 HashAgg 实现，借助落盘来缓解内存压力。
+
+下例通过构造一个占用大量内存的 SQL 语句，对该功能进行演示：
+
+1. 配置 SQL 的 Memory Quota 为 1GB（默认 1GB）。 `set tidb_mem_quota_query = 1 << 30`
+2. 创建单表 `CREATE TABLE t(a int);` 并插入 256 行不同的数据。
+3. 尝试执行 SQL `explain analyze select /*+ HASH_AGG() */ count(*) from t t1 join t t2 join t t3 group by t1.a, t2.a, t3.a;`。该 SQL 占用巨大的内存，会被 tidb-server cancel 掉。
+
```
Comment (Liuxiaozhen12) [inline configure-memory-usage.md:142]: should we add the {{< copyable "sql" >}} here ? PTAL~

File: configure-memory-usage.md (hunk 2 above)
Comment (TomShawn) [reply configure-memory-usage.md]: @wshwsh12 PTAL

File: configure-memory-usage.md [hunk 7]
```diff
@@ -124,3 +124,75 @@ server-memory-quota = 34359738368
...
+### 流控
+
+- TiDB 目前支持对读数据算子的动态内存控制功能。读数据的算子默认启用 [`tidb_disql_scan_concurrency`](/system-variables.md#tidb_distsql_scan_concurrency) 所允许的最大线程数来读取数据。当单条 SQL 语句的内存使用每超过 [`tidb_mem_quota_query`](/system-variables.md#tidb_mem_quota_query) 一次，读数据的算子会停止一个线程。
+- 流控行为会由参数 [`tidb_enable_rate_limit_action`](/system-variables.md#tidb_enable_rate_limit_action) 控制。
+- 当流控被触发时，会在日志中打印一条包含关键字 `memory exceeds quota, destroy one token now` 的日志。
+
+### 落盘
+
+- TiDB 目前支持对执行算子的数据落盘功能。当 SQL 的内存使用超过 Memory Quota 时，tidb-server 可以通过落盘执行算子的中间数据，缓解内存压力。目前支持落盘的算子有：Sort，MergeJoin，HashJoin，HashAgg。
+- 落盘行为会由参数 [`mem-quota-query`](/tidb-configuration-file.md#mem-quota-query)，[`oom-use-tmp-storage`](/tidb-configuration-file.md#oom-use-tmp-storage)，[`oom-use-tmp-storage`](/tidb-configuration-file.md#oom-use-tmp-storage), [`tmp-storage-path`](/tidb-configuration-file.md#tmp-storage-path),[`tmp-storage-quota`](/tidb-configuration-file.md#tmp-storage-quota) 共同控制。
```
Comment (qiancai) [inline configure-memory-usage.md:139]: oom-use-tmp-storage 参数写了两遍

File: configure-memory-usage.md [hunk 8]
```diff
@@ -124,3 +124,75 @@ server-memory-quota = 34359738368
...
+- 当落盘被触发时，会在日志中打印一条包含关键字 `memory exceeds quota, spill to disk now` 或 `memory exceeds quota, set aggregate mode to spill-mode` 的日志。
+- Sort，MergeJoin，HashJoin 落盘是从 v4.0.0 版本开始引入的，HashAgg 落盘是从 v5.2.0 版本开始引入的。
+- 含 HashAgg 的 SQL 引起 OOM 时，可以尝试设置 concurrency = 1 来触发落盘。
```
Comment (qiancai) [inline configure-memory-usage.md:142]: 对于除 HashAgg 外的其他算子，是不是不支持手动设置参数触发落盘呢？

**Review by qiancai — COMMENTED** [review]
@wshwsh12 PTAL

File: configure-memory-usage.md (hunk 8 above)
Comment (wshwsh12) [reply configure-memory-usage.md:142]: 其他算子落盘是默认开启。 可以通过 oom-use-tmp-storage 来控制所有的落盘行为。

**Comment by XuHuaiyu** [comment]
LGTM

File: configure-memory-usage.md [hunk 9]
```diff
@@ -124,3 +124,75 @@ server-memory-quota = 34359738368
...
+    |     └─TableReader_17(Probe)     | 256.00      | 256      | root      |               | time:211.1µs, loops:2, cop_task: {num: 1, max: 295.5µs, proc_keys: 0, rpc_num: 1, rpc_time: 279.7µs, copr_cache_hit_ratio: 0.00}                                    | data:TableFullScan_16                                           | 885 Bytes | N/A      |
+    |       └─TableFullScan_16        | 256.00      | 256      | cop[tikv] | table:t1      | tikv_task:{time:71.4µs, loops:256}                                                                                                                                  | keep order:false, stats:pseudo                                  | N/A       | N/A      |
+    +---------------------------------+-------------+----------+-----------+---------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------+-----------------------------------------------------------------+-----------+----------+
+    9 rows in set (1 min 37.428 sec)
+    ```
+
+> **注意： HashAgg 落盘功能的限制**
```
Comment (TomShawn) [inline configure-memory-usage.md:195]:
```suggestion
> **注意：**
```

File: configure-memory-usage.md [hunk 10]
```diff
@@ -124,3 +124,76 @@ server-memory-quota = 34359738368
...
+
+- TiDB 目前支持对读数据算子的动态内存控制功能。读数据的算子默认启用 [`tidb_disql_scan_concurrency`](/system-variables.md#tidb_distsql_scan_concurrency) 所允许的最大线程数来读取数据。当单条 SQL 语句的内存使用每超过 [`tidb_mem_quota_query`](/system-variables.md#tidb_mem_quota_query) 一次，读数据的算子会停止一个线程。
+- 流控行为会由参数 [`tidb_enable_rate_limit_action`](/system-variables.md#tidb_enable_rate_limit_action) 控制。
+- 当流控被触发时，会在日志中打印一条包含关键字 `memory exceeds quota, destroy one token now` 的日志。
+
+### 落盘
+
+TiDB 支持对执行算子的数据落盘功能。当 SQL 的内存使用超过 Memory Quota 时，tidb-server 可以通过落盘执行算子的中间数据，缓解内存压力。支持落盘的算子有：Sort、MergeJoin、HashJoin、HashAgg。
+
+- 落盘行为会由参数 [`mem-quota-query`](/tidb-configuration-file.md#mem-quota-query)，[`oom-use-tmp-storage`](/tidb-configuration-file.md#oom-use-tmp-storage), [`tmp-storage-path`](/tidb-configuration-file.md#tmp-storage-path),[`tmp-storage-quota`](/tidb-configuration-file.md#tmp-storage-quota) 共同控制。
```
Comment (qiancai) [inline configure-memory-usage.md:140]:
```suggestion
- 落盘行为由参数 [`mem-quota-query`](/tidb-configuration-file.md#mem-quota-query)、[`oom-use-tmp-storage`](/tidb-configuration-file.md#oom-use-tmp-storage)、 [`tmp-storage-path`](/tidb-configuration-file.md#tmp-storage-path)、[`tmp-storage-quota`](/tidb-configuration-file.md#tmp-storage-quota) 共同控制。
```

File: configure-memory-usage.md [hunk 11]
```diff
@@ -124,3 +124,76 @@ server-memory-quota = 34359738368
...
+- 当落盘被触发时，会在日志中打印一条包含关键字 `memory exceeds quota, spill to disk now` 或 `memory exceeds quota, set aggregate mode to spill-mode` 的日志。
```
Comment (qiancai) [inline configure-memory-usage.md:141]:
```suggestion
- 当落盘被触发时，TiDB 会在日志中打印一条包含关键字 `memory exceeds quota, spill to disk now` 或 `memory exceeds quota, set aggregate mode to spill-mode` 的日志。
```

File: configure-memory-usage.md [hunk 12]
```diff
@@ -124,3 +124,75 @@ server-memory-quota = 34359738368
...
+- Sort、MergeJoin、HashJoin 落盘是从 v4.0.0 版本开始引入的，HashAgg 落盘是从 v5.2.0 版本开始引入的。
+- 当包含 Sort、MergeJoin 或 HashJoin 的 SQL 语句引起内存 OOM 时，TiDB 默认会触发落盘。当包含 HashAgg 算子的 SQL 语句引起内存 OOM 时，TiDB 默认不触发落盘，请按照以下示例步骤设置 `tidb_executor_concurrency = 1` 来触发落盘。
```
Comment (qiancai) [inline configure-memory-usage.md:143]:
```suggestion
- 当包含 Sort、MergeJoin 或 HashJoin 的 SQL 语句引起内存 OOM 时，TiDB 默认会触发落盘。当包含 HashAgg 算子的 SQL 语句引起内存 OOM 时，TiDB 默认不触发落盘，请设置系统变量 `tidb_executor_concurrency = 1` 来触发 HashAgg 落盘功能。
```

File: configure-memory-usage.md [hunk 13]
```diff
@@ -124,3 +124,75 @@ server-memory-quota = 34359738368
     * `record path`：表示状态文件存放的目录
 
 5. 通过访问状态文件所在目录（该示例中的目录为 `/tmp/1000_tidb/MC4wLjAuMDo0MDAwLzAuMC4wLjA6MTAwODA=/tmp-storage/record`），可以得到一组文件，其中包括 `goroutinue`、`heap`、`running_sql` 3 个文件，文件以记录状态文件的时间为后缀。这 3 个文件分别用来记录报警时的 goroutine 栈信息，堆内存使用状态，及正在运行的 SQL 信息。其中 `running_sql` 文件内的日志格式请参考 [`expensive-queries`](/identify-expensive-queries.md)。
+
+## tidb-server 其它内存控制策略
+
+### 流控
+
+- TiDB 目前支持对读数据算子的动态内存控制功能。读数据的算子默认启用 [`tidb_disql_scan_concurrency`](/system-variables.md#tidb_distsql_scan_concurrency) 所允许的最大线程数来读取数据。当单条 SQL 语句的内存使用每超过 [`tidb_mem_quota_query`](/system-variables.md#tidb_mem_quota_query) 一次，读数据的算子会停止一个线程。
+- 流控行为会由参数 [`tidb_enable_rate_limit_action`](/system-variables.md#tidb_enable_rate_limit_action) 控制。
```
Comment (qiancai) [inline configure-memory-usage.md:133]:
```suggestion
- 流控行为由参数 [`tidb_enable_rate_limit_action`](/system-variables.md#tidb_enable_rate_limit_action) 控制。
```

File: configure-memory-usage.md [hunk 14]
```diff
@@ -124,3 +124,75 @@ server-memory-quota = 34359738368
...
+
+    {{< copyable "sql" >}}
+
+    ```sql
+    set tidb_mem_quota_query = 1 << 30;
+    ```
+
+2. 创建单表 `CREATE TABLE t(a int);` 并插入 256 行不同的数据。
+
+3. 尝试执行下列 SQL 语句。该 SQL 语句占用大量内存，会返回 Out of Memory Quota 错误，被 tidb-server cancel 掉。
```
Comment (qiancai) [inline configure-memory-usage.md:161]:
```suggestion
3. 尝试执行以下 SQL 语句：
```

File: configure-memory-usage.md [hunk 15]
```diff
@@ -124,3 +124,75 @@ server-memory-quota = 34359738368
...
+
+    {{< copyable "sql" >}}
+
+    ```sql
+    [tidb]> explain analyze select /*+ HASH_AGG() */ count(*) from t t1 join t t2 join t t3 group by t1.a, t2.a, t3.a;
+    ERROR 1105 (HY000): Out Of Memory Quota![conn_id=3]
+    ```
+
+4. 通过用户变量 `tidb_executor_concurrency`，将执行器的并发度调整为 1。在此配置下，内存不足时 HashAgg 会使用尝试触发落盘。
```
Comment (qiancai) [inline configure-memory-usage.md:170]:
```suggestion
4. 设置系统变量 `tidb_executor_concurrency`将执行器的并发度调整为 1。在此配置下，内存不足时 HashAgg 会自动尝试触发落盘。
```

File: configure-memory-usage.md [hunk 16]
```diff
@@ -124,3 +124,75 @@ server-memory-quota = 34359738368
...
+
+2. 创建单表 `CREATE TABLE t(a int);` 并插入 256 行不同的数据。
+
+3. 尝试执行下列 SQL 语句。该 SQL 语句占用大量内存，会返回 Out of Memory Quota 错误，被 tidb-server cancel 掉。
+
+    {{< copyable "sql" >}}
+
+    ```sql
+    [tidb]> explain analyze select /*+ HASH_AGG() */ count(*) from t t1 join t t2 join t t3 group by t1.a, t2.a, t3.a;
+    ERROR 1105 (HY000): Out Of Memory Quota![conn_id=3]
```
Comment (qiancai) [inline configure-memory-usage.md:167]: “ERROR 1105 (HY000): Out Of Memory Quota![conn_id=3]” 这句是输出，请单独写下，不要写到输入里

File: configure-memory-usage.md [hunk 17]
```diff
@@ -124,3 +124,75 @@ server-memory-quota = 34359738368
     * `record path`：表示状态文件存放的目录
 
 5. 通过访问状态文件所在目录（该示例中的目录为 `/tmp/1000_tidb/MC4wLjAuMDo0MDAwLzAuMC4wLjA6MTAwODA=/tmp-storage/record`），可以得到一组文件，其中包括 `goroutinue`、`heap`、`running_sql` 3 个文件，文件以记录状态文件的时间为后缀。这 3 个文件分别用来记录报警时的 goroutine 栈信息，堆内存使用状态，及正在运行的 SQL 信息。其中 `running_sql` 文件内的日志格式请参考 [`expensive-queries`](/identify-expensive-queries.md)。
+
+## tidb-server 其它内存控制策略
+
+### 流控
+
+- TiDB 目前支持对读数据算子的动态内存控制功能。读数据的算子默认启用 [`tidb_disql_scan_concurrency`](/system-variables.md#tidb_distsql_scan_concurrency) 所允许的最大线程数来读取数据。当单条 SQL 语句的内存使用每超过 [`tidb_mem_quota_query`](/system-variables.md#tidb_mem_quota_query) 一次，读数据的算子会停止一个线程。
```
Comment (qiancai) [inline configure-memory-usage.md:132]:
```suggestion
- TiDB 支持对读数据算子的动态内存控制功能。读数据的算子默认启用 [`tidb_disql_scan_concurrency`](/system-variables.md#tidb_distsql_scan_concurrency) 所允许的最大线程数来读取数据。当单条 SQL 语句的内存使用每超过 [`tidb_mem_quota_query`](/system-variables.md#tidb_mem_quota_query) 一次，读数据的算子就会停止一个线程。
```

File: configure-memory-usage.md [hunk 18]
```diff
@@ -124,3 +124,75 @@ server-memory-quota = 34359738368
     * `record path`：表示状态文件存放的目录
 
 5. 通过访问状态文件所在目录（该示例中的目录为 `/tmp/1000_tidb/MC4wLjAuMDo0MDAwLzAuMC4wLjA6MTAwODA=/tmp-storage/record`），可以得到一组文件，其中包括 `goroutinue`、`heap`、`running_sql` 3 个文件，文件以记录状态文件的时间为后缀。这 3 个文件分别用来记录报警时的 goroutine 栈信息，堆内存使用状态，及正在运行的 SQL 信息。其中 `running_sql` 文件内的日志格式请参考 [`expensive-queries`](/identify-expensive-queries.md)。
+
+## tidb-server 其它内存控制策略
+
+### 流控
```
Comment (qiancai) [inline configure-memory-usage.md:130]: 这里要不要写成 “动态内存控制功能“

File: configure-memory-usage.md [hunk 19]
```diff
@@ -124,3 +124,75 @@ server-memory-quota = 34359738368
...
+
+- TiDB 目前支持对读数据算子的动态内存控制功能。读数据的算子默认启用 [`tidb_disql_scan_concurrency`](/system-variables.md#tidb_distsql_scan_concurrency) 所允许的最大线程数来读取数据。当单条 SQL 语句的内存使用每超过 [`tidb_mem_quota_query`](/system-variables.md#tidb_mem_quota_query) 一次，读数据的算子会停止一个线程。
+- 流控行为会由参数 [`tidb_enable_rate_limit_action`](/system-variables.md#tidb_enable_rate_limit_action) 控制。
+- 当流控被触发时，会在日志中打印一条包含关键字 `memory exceeds quota, destroy one token now` 的日志。
+
+### 落盘
```
Comment (qiancai) [inline configure-memory-usage.md:136]: 这里要不要写“数据落盘”

File: configure-memory-usage.md (hunk 18 above)
Comment (wshwsh12) [reply configure-memory-usage.md]: 写成”流量控制“ 可以吗？    ”动态内存控制功能“太宽泛，数据落盘也能算动态内存控制功能

File: configure-memory-usage.md [hunk 20]
```diff
@@ -124,3 +124,81 @@ server-memory-quota = 34359738368
...
+2. 创建单表 `CREATE TABLE t(a int);` 并插入 256 行不同的数据。
+
+3. 尝试执行以下 SQL 语句：
+
+    {{< copyable "sql" >}}
+
+    ```sql
+    [tidb]> explain analyze select /*+ HASH_AGG() */ count(*) from t t1 join t t2 join t t3 group by t1.a, t2.a, t3.a;
+    ```
+
```
Comment (qiancai) [inline configure-memory-usage.md:168]:
```suggestion

该 SQL 语句占用大量内存，返回 Out of Memory Quota 错误。

```

**Review by qiancai — APPROVED** [review]
LGTM

**Review by XuHuaiyu — COMMENTED** [review]
LGTM

**Review by XuHuaiyu — APPROVED** [review]

**Pull request merged**